packages/core/src/browser/label-provider.ts

```diff
@@ -114,52 +120,24 @@ export class LabelProvider implements FrontendApplicationContribution {
      */
     initialize(): void {
         const contributions = this.contributionProvider.getContributions();
-        for (const contribution of contributions) {
-            if (contribution.onDidChange) {
-                contribution.onDidChange(event => {
-                    const affects = (uri: URI) => this.affects(uri, event, contribution);
-                    this.onDidChangeEmitter.fire({ affects });
+        for (const eventContribution of contributions) {
+            if (eventContribution.onDidChange) {
+                eventContribution.onDidChange(event => {
+                    this.onDidChangeEmitter.fire({
+                        affects: element => this.affects(element, event)
+                    });
                 });
             }
         }
     }
 
-    /**
-     * When the given event occurs, determine if the given URI could in any
-     * way be affected.
-     *
-     * If the event directly indicates that it affects the URI then of course we
-     * return `true`.  However there may be label provider contributions that delegate
-     * back to the label provider.  These contributors do not, and should not, listen for
-     * label provider events because that would cause infinite recursion.
-     *
-     * @param uri
-     * @param event
-     */
-    protected affects(element: object, event: DidChangeLabelEvent, originatingContribution: LabelProviderContribution): boolean {
-
-        const contribs = this.findContribution(element);
-        const possibleContribsWithDups = [
-            contribs.find(c => c.getIcon !== undefined),
-            contribs.find(c => c.getName !== undefined),
-            contribs.find(c => c.getLongName !== undefined),
-        ];
-        const possibleContribsWithoutDups = [...new Set(possibleContribsWithDups)];
-        for (const possibleContrib of possibleContribsWithoutDups) {
-            if (possibleContrib) {
-                if (possibleContrib === originatingContribution) {
-                    if (event.affects(element)) {
-                        return true;
-                    }
-                }
-                if (possibleContrib.getConstituentUris) {
-                    const constituentUris: URI[] = possibleContrib.getConstituentUris(element);
-                    for (const constituentUri of constituentUris) {
-                        if (this.affects(constituentUri, event, originatingContribution)) {
-                            return true;
-                        }
-                    }
-                }
+    protected affects(element: object, event: DidChangeLabelEvent): boolean {
+        if (event.affects(element)) {
+            return true;
+        }
+        for (const contribution of this.findContribution(element)) {
+            if (contribution.affects && contribution.affects(element, event)) {
+                return true;
```

Comment:
I don't think that the code before was correct, i.e an order given by `canHandle` should be used. We should not sort to get some other order. Also I don't get how it is possible to not fire an event from second contribution, for some elements it can return a value for other not. It should be guarded by `canHandle` and did change event implementations. If a client then calls `affects` it can decide to skip refreshing itself or can always do it if it is cheap.